python/cudf_polars/cudf_polars/experimental/distinct.py

```diff
@@ -0,0 +1,149 @@
+from cudf_polars.dsl.ir import Distinct
+from cudf_polars.experimental.base import PartitionInfo
+from cudf_polars.experimental.dispatch import lower_ir_node
+from cudf_polars.experimental.utils import _lower_ir_fallback
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.dispatch import LowerIRTransformer
+    from cudf_polars.utils.config import ConfigOptions
+
+
+def lower_distinct(
+    ir: Distinct,
+    child: IR,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    config_options: ConfigOptions,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    """
+    Lower a Distinct IR into partition-wise stages.
+
+    Parameters
+    ----------
+    ir
+        The Distinct IR node to lower.
+    child
+        The reconstructed child of ``ir``. May differ
+        from ``ir.children[0]``.
+    partition_info
+        A mapping from all unique IR nodes to the
+        associated partitioning information.
+    config_options
+        GPUEngine configuration options.
+
+    Returns
+    -------
+    new_node
+        The lowered node (usually Distinct).
+    partition_info
+        A mapping from unique nodes in the new graph to associated
+        partitioning information.
+    """
+    from cudf_polars.experimental.repartition import Repartition
+    from cudf_polars.experimental.shuffle import Shuffle
+
+    # Extract child partitioning
+    child_count = partition_info[child].count
+
+    # Assume shuffle is not stable for now
+    require_tree_reduction = ir.stable or ir.keep in (
+        plc.stream_compaction.DuplicateKeepOption.KEEP_FIRST,
+        plc.stream_compaction.DuplicateKeepOption.KEEP_LAST,
+    )
+
+    subset: frozenset = ir.subset or frozenset(ir.schema)
+    shuffle_keys = tuple(NamedExpr(name, Col(ir.schema[name], name)) for name in subset)
+    shuffled = partition_info[child].partitioned_on == shuffle_keys
+    if ir.keep == plc.stream_compaction.DuplicateKeepOption.KEEP_NONE:
+        # Need to shuffle the original data for keep == "none"
+        if require_tree_reduction:
+            raise NotImplementedError(
+                "Unsupported unique options for multiple partitions."
+            )
+        if not shuffled:
+            child = Shuffle(child.schema, shuffle_keys, config_options, child)
+            partition_info[child] = PartitionInfo(
+                count=child_count,
+                partitioned_on=shuffle_keys,
+            )
+            shuffled = True
+
+    output_count = 1
+    n_ary = 32  # Arbitrary default (for now)
+    if ir.zlice is not None:
+        # Head/tail slice operation has been pushed into Distinct
+        if ir.zlice[0] < 1 and ir.zlice[1] is not None:
+            # Use rough 1m-row heuristic to set n_ary
+            n_ary = max(int(1_000_000 / ir.zlice[1]), 2)
+        else:  # pragma: no cover
+            # Slice is not head or tail - Can this happen?
```

Comment:
Is that right?
```suggestion
            # TODO: Support slice pushdown more generally
```

Reply:
I *think* that is right.  However, I doubt we ever want to support an arbitrary slice, because we would need to calculate the length of every partition to get the offsets.